keps/sig-storage/20200116-node-failover.md

```diff
@@ -0,0 +1,178 @@
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+In case of an instance being shutdown (i.e. not running) the control plane doesn't make the right
+assumptions to enable stateful workload to fail-over, This KEP introduces a flow to do so.
+
+This KEP is part of collection of KEPs that intends to enhance stateful workload management on top of Kubernetes.
+
+## Motivation
+
+Today kubernetes doesn't enable providers that are able to safely detach (i.e. only detach when volumes aren't being written to) to do so. This proposal aims to introduce a new CSI capability that if advertised would trigger volume detach if a node is partitioned from the control plane.
+
+### Goals
+
+* Increase the availability of stateful workloads
+* Automate self-healing for stateful workloads
+
+### Non-Goals
+
+* Implement in-cluster logic to handle node/control plane partitioning 
+* enable detach for all the storage providers
+
+
+## Proposal
+
+user stories:
+
+* as a cluster administrator I want my stateful workload to failover in case of a node is partitioned without any intervention when it's safe to do so
+
+* as a developer I would like to rely on a self-healing platform for my stateful workloads
+
+This KEP plans to introduce the following capability to the CSI capability to the spec:
+
+```go
+// CSIDriverSpec is the specification of a CSIDriver.
+type CSIDriverSpec struct {
+  ...
+
+	// SafeDetach indicates this CSI volume driver is able to perform detach
+  // operations, when set to true, the CSI volume driver needs to ensure that 
+  // ControllerUnpublishVolume() implementation is checking for detach safety
+	// +optional
+  SafeDetach *bool
+}
+
+```
+
+
+the kube-controller-manager, would need to ensure that eviction looks at the `CSIDriver` and when `.spec.SafeDetach` is set true, forcefully delete pods.
+
+Specifically the pod gc controller (the only controller able to forcefully delete pods) would also check now for stateful pods, and select those that satisfy ALL of the following conditions:
+
+- pods that are backed by a provisioner that support safe detach
+- pods that are backed by volumes that aren't marked as `ReadWriteMany, this would avoid breaking having multiple replicas of the same pod.
+
+once selected and forcefully deleted, thi should lead to the deletion of the `volumeAttachement` object, which would in turn trigger `ControllerUnpublishVolume()`, the provider would need to ensure there that no volume is being used through a mount point. If unable to determine the usage, the storage provider can return an error without detaching the volume, the `external-attacher` would then retry with backoff.
```

Comment:
I did some experiment with this.

1. Created a statefulset with 3 replicas.
2. Shut down 1 node. The pod on that node is stuck in Terminating state. PVC/PV and VolumeAttachment stay the same.
3. Forcefully delete the pod, setting grace-period=0. The old pod is deleted, but a new pod is stuck in ContainerCreating state because the old VolumeAttachment is not deleted.
4. Manually delete the old VolumeAttachment.
5. Now the new pod is running on a different node.

So line 122 ("once selected and forcefully deleted, this should lead to the deletion of the `volumeAttachement` object") needs an update.

Reply:
fixed